dtm/clk_src.h

```diff
@@ -0,0 +1,162 @@
+ *
+ *   Clock source module is designed to be used by DTM0. At this point,
+ * only one type of clock is supported -- physical clock. Such a clock
+ * provides a way to enforce total ordering in a system with synchronised
+ * clocks. However, the interface supports an epoch-based clock source as well.
+ *   See "A.clock.sync" in the DTM0 HLD for additional details.
+ *
+ * Interface
+ * ---------
+ *
+ * CS.TS          -- A data type that represents a point in time (timestamp).
+ * CS.TS.CMP(L,R) -- a 3-way comparison for timestamps; it returns
+ *                   -1 when "left" happened before "right", +1 when
+ *                   "left" happened after "right", and 0 otherwise.
+ * CS.INIT        -- initialise a clock source of the given type.
+ * CS.FINI        -- finalise a lock source.
+ * CS.NOW         -- get the current value of a clock source.
+ * CS.OBSERVED(O) -- make the clock source aware of a timestamp (Other)
+ *                   generated by another instance of the same clock source
+ *                   type.
+ *
+ * Error handling
+ * --------------
+ *
+ *   It is safe to assume that CS.NOW returns "unrecoverable" errors (the caller
+ * should panic if a non-zero value is returned).
+ *   When causality is tracked only by a physical clock (CS_PHYS),
+ * CS.OBSERVED returns -EINVAL if the incomming message "happened" in future.
+ * The caller should stop communication with the clock source that has
+ * generated such a timestamp (notify HA or, at least, panic right here).
+ *
+ * Concurrency
+ * -----------
+ *
+ *   It is safe to call CS.NOW and CS.OBSERVED from any context as long
+ * as the caller owns the corresponding source clock.
+ *   The user must ensure to use only one type of clock source in the system.
+ * The module provides no protection agains such cases.
+ */
+
+
+/**
+ * @defgroup dtm
+ *
+ * @{
+ */
+
+
+enum m0_dtm_cs_types {
+	/* Physical clock.
+	 * Physical clock supports total ordering and requires
+	 * no synchronisation (from DTM0 perspective). But it has
+	 * two drawbacks: simultaneous events are not allowed,
+	 * and the current clock source value cannot be less
+	 * than the timestamp of any event in the system delivered
+	 * to the node.
+	 */
+	M0_DTM_CS_PHYS,
+};
+
+/** A data type that represents a timestamp (see CS.TS).*/
+struct m0_dtm_ts {
+	m0_time_t dts_phys;
+} M0_XCA_RECORD M0_XCA_DOMAIN(rpc);
+
+/** Defines the minimal valid value for a CS.TS. */
+#define M0_DTM_TS_MIN (struct m0_dtm_ts) { .dts_phys = 1 }
+/** Defines the maximal valid value for a CS.TS. */
+#define M0_DTM_TS_MAX (struct m0_dtm_ts) { .dts_phys = (UINT64_MAX - 1) }
+/** Defines an invalid (but defined) value for a CS.TS. */
+#define M0_DTM_TS_INIT (struct m0_dtm_ts) { .dts_phys = UINT64_MAX }
+
+#define M0_DTM_TS_P(_ts) ((_ts)->dts_phys)
+#define M0_DTM_TS_FMT "@%" PRIu64
+
+/** Timestamp ordering is the result of comparison of two timestamps
+ * (see CS.TS.CMP for the details).
+ */
+enum m0_dtm_ts_ord {
```

Comment:
do you really need an explicit enum for this? +-1/0 is a known convention for ordering.

Reply:
I agree with the last point (it is a known convention) but from readability standpoint it is much easier to use explicit LT,EQ,GT symbols rather than the bare literals (-1,0,1). When it comes to timestamp relation, these literals require two "jumps" inside the reader's head (-1 -> less -> happened-before) while enum requires only one (LT -> happened-before).
However, if you are not convinced by this explanation, please let me know -- I'll replace the enum with bare integerl literals.